Document IoT common queries (#3013)
influxdata/docs-v2: content/influxdb/v2.0/query-data/common-queries/operate-on-columns.md

---
title: Operate on columns
description: >
  Find and count unique values, recalculate the `_value` column, and use values to calculate a new column.
influxdb/v2.0/tags: [queries]
aliases: 
  - /influxdb/cloud/query-data/common-queries/count_unique_values_for_column/
  - /influxdb/cloud/query-data/common-queries/recalculate_value_column/
  - /influxdb/cloud/query-data/common-queries/calculate_new_column/
menu:
  influxdb_2_0:
    name: Operate on columns
    parent: Common queries
weight: 100  
---

Use the following common queries to operate on columns:

- [Find and count unique values in a column](#find-and-count-unique-values-in-a-column)
- [Recalculate the _values column](#recalculate-the-_values-column)
- [Calculate a new column](#calculate-a-new-column)

{{% note %}}
These examples use [NOAA water sample data](/influxdb/v2.0/reference/sample-data/#noaa-water-sample-data).
{{% /note %}}

## Find and count unique values in a column

Count the number of unique values in a specified column.

The following examples identify and count unique locations that data was collected from.

### Find unique values

This query:

>   - Uses [`group()`](/influxdb/v2.0/reference/flux/stdlib/built-in/transformations/group/) to ungroup data and return results in a single table.
>   - Uses [`keep()`](/influxdb/v2.0/reference/flux/stdlib/built-in/transformations/keep/) and [`unique()`](/influxdb/v2.0/reference/flux/stdlib/built-in/transformations/selectors/unique/) to return unique values in the specified column.

Reviewer: ```suggestion
  - Uses [`group()`](/{{< latest "flux" >}}/stdlib/universe/group/) to ungroup data and return results in a single table.
  - Uses [`keep()`](/{{< latest "flux" >}}/stdlib/universe/keep/) and [`unique()`](/{{< latest "flux" >}}/stdlib/universe/selectors/unique/) to return unique values in the specified column.
```

```js
from(bucket: "noaa")
  |> range(start: -30d)
  |> group()
  |> keep(columns: ["location"])
  |> unique(column: "location")
```

#### Example results
| location     |
|:--------     |
| coyote_creek |
| santa_monica |

### Count unique values

This query:

> - Uses [`group()`](/influxdb/v2.0/reference/flux/stdlib/built-in/transformations/group/) to ungroup data and return results in a single table.
> - Uses [`keep()`](/influxdb/v2.0/reference/flux/stdlib/built-in/transformations/keep/), [`unique()`](/influxdb/v2.0/reference/flux/stdlib/built-in/transformations/selectors/unique/), and then [`count()`](/influxdb/v2.0/reference/flux/stdlib/built-in/transformations/aggregates/count/) to count the number of unique values.

Reviewer: ```suggestion
- Uses [`group()`](/{{< latest "flux" >}}/stdlib/universe/group/) to ungroup data and return results in a single table.
- Uses [`keep()`](/{{< latest "flux" >}}/stdlib/universe/keep/), [`unique()`](/{{< latest "flux" >}}/stdlib/universe/unique/), and then [`count()`](/{{< latest "flux" >}}/stdlib/universe/count/) to count the number of unique values.
```

```js
from(bucket: "noaa")
  |> group()
  |> unique(column: "location")
  |> count(column: "location")
```

#### Example results

| location  |
| ---------:|
| 2         |

> Use values to calculate a new column.
> 
>  Use the `map()` function to create a new column calculated from existing values in each row.

Reviewer: ```suggestion
```

>  Use the `map()` function to create a new column calculated from existing values in each row.

Reviewer: These seem out of place.

## Recalculate the _values column

> Recalculate the `_value` column without creating a new one. Use the `with` operator in [`map()`](/{{< latest "flux" >}}/stdlib/universe/map/) to overwrite the existing `_value` column.

Reviewer: ```suggestion
To recalculate the `_value` column, use the `with` operator in [`map()`](/{{< latest "flux" >}}/stdlib/universe/map/) to overwrite the existing `_value` column.
```

> Recalculate the `_value` column without creating a new one. Use the `with` operator in [`map()`](/influxdb/v2.0/reference/flux/stdlib/built-in/transformations/map/) to overwrite the existing `_value` column.

Reviewer: ```suggestion
Recalculate the `_value` column without creating a new one. Use the `with` operator in [`map()`](/{{< latest "flux" >}}/stdlib/universe/map/) to overwrite the existing `_value` column.
```

The following query:

>   - Uses [`filter()`](/influxdb/v2.0/reference/flux/stdlib/built-in/transformations/filter/) to filter the `average_temperature` measurement.
>   - Uses [`map()`](/influxdb/v2.0/reference/flux/stdlib/built-in/transformations/map/) to convert Fahrenheit temperature values into Celsius.

Reviewer: ```suggestion
  - Uses [`filter()`](/{{< latest "flux" >>}}/stdlib/universe/filter/) to filter the `average_temperature` measurement.
  - Uses [`map()`](/{{< latest "flux" >>}}/stdlib/universe/map/) to convert Fahrenheit temperature values into Celsius.
```

| degrees | average_temperature | 1920-03-05T22:10:01Z | 2020-03-05T22:10:01Z | 2019-08-17T00:24:00Z | coyote_creek | 25                 |
| degrees | average_temperature | 1920-03-05T22:10:01Z | 2020-03-05T22:10:01Z | 2019-08-17T00:30:00Z | coyote_creek | 21.11111111111111  |
| degrees | average_temperature | 1920-03-05T22:10:01Z | 2020-03-05T22:10:01Z | 2019-08-17T00:36:00Z | coyote_creek | 28.88888888888889  |
| degrees | average_temperature | 1920-03-05T22:10:01Z | 2020-03-05T22:10:01Z | 2019-08-17T00:42:00Z | coyote_creek | 24.444444444444443 |
| degrees | average_temperature | 1920-03-05T22:10:01Z | 2020-03-05T22:10:01Z | 2019-08-17T00:48:00Z | coyote_creek | 29.444444444444443 |
| degrees | average_temperature | 1920-03-05T22:10:01Z | 2020-03-05T22:10:01Z | 2019-08-17T00:54:00Z | coyote_creek | 26.666666666666668 |
| degrees | average_temperature | 1920-03-05T22:10:01Z | 2020-03-05T22:10:01Z | 2019-08-17T01:00:00Z | coyote_creek | 21.11111111111111  |
| •••     | •••                 | •••                  | •••                  | •••                  | •••          | •••                |

## Calculate a new column

> Use the `map()` function to create a new column calculated from existing values in each row.
> 
> This example converts temperature from Fahrenheit to Celsius and maps the Celsius value to a new `celsius` column.

Reviewer: ```suggestion
To use values in a row to calculate and add a new column, use `map()`.
This example below converts temperature from Fahrenheit to Celsius and maps the Celsius value to a new `celsius` column.
```

The following query:

>   - Uses [`filter()`](/influxdb/v2.0/reference/flux/stdlib/built-in/transformations/filter/) to filter the `average_temperature` measurement.
>   - Uses [`map()`](/influxdb/v2.0/reference/flux/stdlib/built-in/transformations/map/) to create a new column calculated from existing values in each row.

Reviewer: ```suggestion
  - Uses [`filter()`](/{{< latest "flux" >}}/stdlib/universe/filter/) to filter the `average_temperature` measurement.
  - Uses [`map()`](/{{< latest "flux" >}}/stdlib/universe/map/) to create a new column calculated from existing values in each row.
```
